File: kmp.py
from typing import *

def calculateLps(part:  str) -> List[int]:
    i,j = 1,0
    lps = [0]*len(part)

    while i < len(part):
        if part[i]==part[j]:
            lps[i] = lps[j] + 1
            i += 1
            j += 1
        elif j == 0:
            lps[i] = 0
            i += 1
        else:
            j = lps[j-1]

    return lps

def kmp(s: str, part: str) -> List[int]:
    i, parti = 0,0
    lps = calculateLps(part)
    res = []    
    while i < len(s):
        if s[i] == part[parti]:
            i += 1
            parti += 1
            if parti == len(part):
                res.append(i - len(part))
                parti = lps[parti - 1]
        else:
            if parti == 0:
                i += 1
            else:
                parti = lps[parti - 1]

    return res

if __name__ == '__main__':
    part ='kpygkivtlqoocssnextkqzjpycbylkaonds'
    part = 'AAACAAAA'
    s = 'aabaacaadaabaaba'
    part = 'aaba'
    lps = calculateLps(part)
    print(lps)
    matches = kmp(s, part)
    print("matches=", matches)

File: kmp.go
package main


import "fmt"

func calculateLps(part string) []int {
    lps := make([]int, len(part))
    i,j := 1,0
    for ; i < len(part) ; {
        if part[i] == part[j] {
            lps[i] = lps[j] + 1
            i +=1
            j +=1
        } else if j == 0 {
            lps[i] = 0
            i += 1
        } else {
            j = lps[j-1]
        }
    }

    return lps
}

func kmp(s string, part string) []int {
    matches := make([]int,0)
    i,j := 0,0
    lps := calculateLps(part)
    
    for ; i < len(s) ; {
        if s[i] == part[j]{
            i += 1
            j += 1
            if j == len(part) {
                idx := i - len(part)
                matches = append(matches, idx)
                j = lps[j-1]
            }
        } else if lps[j] == 0 {
            i += 1
        } else {
            j = lps[j-1]
        }
    }
    
    return matches
}

func main() {
    s := "kpygkivtlqoocskpygkpygkivtlqoocssnextkqzjpycbylkaondskivtlqoocssnextkqzjpycbylkaondssnextkqzjpycbylkaondshijzgaovndkjiiuwjtcpdpbkrfsi"
    part := "kpygkivtlqoocssnextkqzjpycbylkaonds"
    lps := calculateLps(part)
    fmt.Println("lps=",lps)
    matches := kmp(s, part)
    fmt.Println("matches=", matches)
}

